### PR TITLE
Add protections against spams and a configuration page.

### PR DESCRIPTION
**Pull request by alegout**
Add protections against spams and a configuration page.

Ticket: [BOOM-4288](http://forge.prestashop.com/browse/BOOM-4288)

QA: Needs to be tested with 1.7.2 and 1.7.3. Wordings will have to stay in English until 1.7.4

### CODE REVIEW

File: contactform.php [hunk 1]
```diff
@@ -41,29 +53,187 @@ public function __construct()
...
 
+    /**
+     * @return string
+     */
+    public function getContent()
+    {
+
+        $message = $this->trans(
+            'For even more security on your website forms, consult our Security & Access modules category on the %s',
+            array($this->getSecurityMarketPlaceLink()),
```
Comment (Quetzacoalt91) [inline contactform.php]: Don't you want to add placeholders like in the core? Like `%link%`

File: contactform.php [hunk 2]
```diff
@@ -158,15 +349,23 @@ public function getTemplateVarOrders()
...
         $message = Tools::getValue('message');
+        $url = Tools::getValue('url');
+        $clientToken = Tools::getValue('token');
+        $serverToken = $this->context->cookie->contactFormToken;
+        $clientTokenTTL = $this->context->cookie->contactFormTokenTTL;
 
         if (!($from = trim(Tools::getValue('from'))) || !Validate::isEmail($from)) {
             $this->context->controller->errors[] = $this->trans('Invalid email address.', array(), 'Shop.Notifications.Error');
-        } elseif (preg_match('#^[ ]*$#', $message)) {
+        } elseif (!$message) {
```
Comment (Quetzacoalt91) [inline contactform.php]: It seems a message filled with blanks could pass now, should not we test that with a `trim()` and `empty()`

File: contactform.php (hunk 1 above)
Comment (alegout) [reply contactform.php]: totally, thanks :)

File: contactform.php (hunk 2 above)
Comment (alegout) [reply contactform.php]: go for empty 👍

File: contactform.php [hunk 3]
```diff
@@ -41,29 +53,187 @@ public function __construct()
...
     }
 
+    /**
+     * @return string
+     */
+    public function getContent()
+    {
+
+        $message = $this->trans(
+            'For even more security on your website forms, consult our Security & Access modules category on the %s',
```
Comment (eternoendless) [inline contactform.php]: Maybe change "website" to "shop's"?
ping @LouiseBonnard @colinegin

File: contactform.php [hunk 4]
```diff
@@ -41,29 +53,187 @@ public function __construct()
...
+            'form' => array(
+                'legend' => array(
+                    'title' => $this->trans('Parameters', array(), 'Modules.Contactform.Admin'),
+                    'icon' => 'icon-envelope'
+                ),
+                'input' => array(
+                    array(
+                        'type' => 'switch',
+                        'label' => $this->trans('Send confirmation email to your customers', array(), 'Modules.Contactform.Admin'),
+                        'desc' => $this->trans('Click Yes and your customers will receive a confirmation email without the message content', array(), 'Modules.Contactform.Admin'),
```
Comment (eternoendless) [inline contactform.php]: I think this wording could be improved to better guide the merchants:

> Choose Yes and your customers will receive a generic confirmation email including a tracking number after their message is sent. Note: to discourage spam, the content of their message won't be included in the email.

What do you think @LouiseBonnard @colinegin ?

File: contactform.php [hunk 5]
```diff
@@ -41,29 +53,187 @@ public function __construct()
...
+                    'title' => $this->trans('Save', array(), 'Admin.Actions'),
+                )
+            ),
+        );
+        $helper = new HelperForm();
+        $helper->table = $this->table;
+        $lang = new Language((int) Configuration::get('PS_LANG_DEFAULT'));
+        $helper->default_form_language = $lang->id;
+        $helper->submit_action = 'update-configuration';
+        $helper->currentIndex = $this->context->link->getAdminLink('AdminModules', false).'&configure='
```
Comment (eternoendless) [inline contactform.php]: You should use `parse_url()` to make sure that there's a query string in the URL returned by `$this->context->link->getAdminLink('AdminModules', false)`, then `http_build_query()` to rebuild the query string with your arguments.

File: contactform.php [hunk 6]
```diff
@@ -176,14 +375,17 @@ public function sendMessage()
             $this->context->controller->errors[] = $this->trans('An error occurred during the file-upload process.', array(), 'Modules.Contactform.Shop');
         } elseif (!empty($file_attachment['name']) && !in_array(Tools::strtolower(substr($file_attachment['name'], -4)), $extension) && !in_array(Tools::strtolower(substr($file_attachment['name'], -5)), $extension)) {
             $this->context->controller->errors[] = $this->trans('Bad file extension', array(), 'Modules.Contactform.Shop');
+        } elseif ($url !== '' || empty($serverToken)
+            || $clientToken !== $serverToken || $clientTokenTTL < time()
+        ) {
+            $this->context->controller->errors[] = $this->trans('An error occurred while sending the message.', array(), 'Modules.Contactform.Shop');
```
Comment (eternoendless) [inline contactform.php]: How about adding a "Please try again"?

File: contactform.php [hunk 7]
```diff
@@ -176,14 +375,17 @@ public function sendMessage()
             $this->context->controller->errors[] = $this->trans('An error occurred during the file-upload process.', array(), 'Modules.Contactform.Shop');
         } elseif (!empty($file_attachment['name']) && !in_array(Tools::strtolower(substr($file_attachment['name'], -4)), $extension) && !in_array(Tools::strtolower(substr($file_attachment['name'], -5)), $extension)) {
             $this->context->controller->errors[] = $this->trans('Bad file extension', array(), 'Modules.Contactform.Shop');
+        } elseif ($url !== '' || empty($serverToken)
+            || $clientToken !== $serverToken || $clientTokenTTL < time()
```
Comment (eternoendless) [inline contactform.php]: Either one line for all statements or one statement per line please

File: contactform.php (hunk 5 above)
Comment (alegout) [reply contactform.php]: not my code...

File: contactform.php (hunk 6 above)
Comment (alegout) [reply contactform.php]: done

File: contactform.php (hunk 7 above)
Comment (alegout) [reply contactform.php]: done

File: contactform.php (hunk 4 above)
Comment (LouiseBonnard) [reply contactform.php]: Ah oui !

File: contactform.php (hunk 5 above)
Comment (alegout) [reply contactform.php]: done

File: contactform.php (hunk 4 above)
Comment (alegout) [reply contactform.php]: done

**Comment by eternoendless** [comment]
Thank you @alegout

**Pull request merged**